### PR TITLE
Add an option to preserve proto names in JsonFormatter

### LINKED ISSUE #6299: C# JsonFormatter does not have a preservingProtoFieldNames method
**What language does this apply to?**
If it's a proto syntax change, is it for proto2 or proto3?
If it's about generated code change, what programming language?
C#, Proto3

**Describe the problem you are trying to solve.**
The `JsonFormatter` does not have a `preservingProtoFieldNames` method like Java; see [JsonFormat.Printer](https://developers.google.com/protocol-buffers/docs/reference/java/com/google/protobuf/util/JsonFormat.Printer#preservingProtoFieldNames--)

**Describe the solution you'd like**
Adding the method, like in Java; simliar issue #2029

**Describe alternatives you've considered**
None

**Additional context**
Example:
```proto
message FullName {
  string first_name = 1;
  string last_name = 2;
}
```
```csharp
var name = new FullName {FirstName = "John", LastName = "Doe"};
var jsonName = JsonFormatter.Default.Format(name);
Console.Out.WriteLine(jsonName);
```

Output is:
```text
{ "firstName": "John", "lastName": "Doe" }
```

### PR DESCRIPTION
**Pull request by ObsidianMinor**
Fixes #6299

### CODE REVIEW

**Comment by ObsidianMinor** [comment]
This has been sitting here for quite a bit. Any updates? @anandolee

**Comment by ahmelsayed** [comment]
I was looking for this option as well. Any updates on merging this PR?

**Comment by ObsidianMinor** [comment]
Could you also review this as well @jtattermusch?

**Comment by ahmelsayed** [comment]
@anandolee Any updates on this PR?

**Review by anandolee — APPROVED** [review]

**Comment by jtattermusch** [comment]
@jskeet  @amanda-tarafa this probably deserves to be looked at by a pair of C# expert's eyes - just to make sure nothing was missed (I'll try to take a look as well but can't guarantee).

File: csharp/src/Google.Protobuf.Test/JsonFormatterTest.cs
Comment (jtattermusch) [inline csharp/src/Google.Protobuf.Test/JsonFormatterTest.cs:1]: what has changes about this file that it shows up in the diff?
Unless I missed something, the entire file showing up as modified makes it impossible to see any actual changes to the file (which makes reviewing it very difficult).

Let's avoid "false diffs" like this in the future.

File: csharp/src/Google.Protobuf.Test/JsonFormatterTest.cs
Comment (amanda-tarafa) [inline csharp/src/Google.Protobuf.Test/JsonFormatterTest.cs:666]: I think this is the only test that has been added.

**Review by amanda-tarafa — COMMENTED** [review]
I've done a quick review as well. It looks good.

But I'm wondering about the usefulness of this without the counterpart change in JsonParser. But there might be something I'm missing here.

File: csharp/src/Google.Protobuf.Test/JsonFormatterTest.cs
Comment (jskeet) [reply csharp/src/Google.Protobuf.Test/JsonFormatterTest.cs:1]: I believe it's whitespace-only - quite possibly line endings. I'd really like to see the whitespace-only changes here reverted before we merge; it makes it much hard to trace changes.

File: csharp/src/Google.Protobuf/JsonFormatter.cs
Comment (jskeet) [inline csharp/src/Google.Protobuf/JsonFormatter.cs:236]: I think I'd do this with a conditional operator instead:

```csharp
WriteString(writer, settings.PreserveProtoFieldNames ? accessor.Descriptor.Name : accessor.Descriptor.JsonName);
```

File: csharp/src/Google.Protobuf/JsonFormatter.cs
Comment (jskeet) [inline csharp/src/Google.Protobuf/JsonFormatter.cs:830]: Nit: please remove redundant blank line.

File: csharp/src/Google.Protobuf/JsonFormatter.cs
Comment (jskeet) [inline csharp/src/Google.Protobuf/JsonFormatter.cs:890]: Hmm... I *thought* we allowed users to specify a different JSON field name to the one that would be inferred by the compiler otherwise. Not sure - but arguably using "false to use the default JSON field name representation" would cover that either way.

**Review by jskeet — COMMENTED** [review]
It looks like we don't need an option for this in the parser as we already handle using either the JSON field name or the original name (see https://github.com/protocolbuffers/protobuf/blob/b926a7d2097405fd3be4cc4009d61710562852d7/csharp/src/Google.Protobuf/Reflection/MessageDescriptor.cs#L121).

However, I'm concerned with per-language feature creep. #6299 observes that the option *does* exist in Java, but I don't know about any other languages. I think the Protobuf team should decide whether they're happy for this to be a feature in general. If it's approved in general, then this implementation looks broadly okay.

**Comment by jskeet** [comment]
Eek - I hadn't seen that this had been merged as-is.

**Pull request merged**
